swarm_dial.go

```diff
@@ -281,162 +280,375 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
+				continue loop
+			}
+
+			// it must be an error -- add backoff if applicable and dispatch
+			if res.Err != context.Canceled && !connected {
+				// we only add backoff if there has not been a successful connection
+				// for consistency with the old dialer behavior.
+				s.backf.AddBackoff(p, res.Addr)
+			}
+
+			dispatchError(ad, res.Err)
+			if active == 0 && len(nextDial) > 0 {
+				triggerDial = triggerNow
+			}
+		}
 	}
+}
 
-	//////
+func (s *Swarm) addrsForDial(ctx context.Context, p peer.ID) ([]ma.Multiaddr, error) {
 	peerAddrs := s.peers.Addrs(p)
 	if len(peerAddrs) == 0 {
-		return nil, &DialError{Peer: p, Cause: ErrNoAddresses}
+		return nil, ErrNoAddresses
 	}
+
 	goodAddrs := s.filterKnownUndialables(p, peerAddrs)
-	if forceDirect {
+	if forceDirect, _ := network.GetForceDirectDial(ctx); forceDirect {
 		goodAddrs = addrutil.FilterAddrs(goodAddrs, s.nonProxyAddr)
 	}
+
 	if len(goodAddrs) == 0 {
-		return nil, &DialError{Peer: p, Cause: ErrNoGoodAddresses}
+		return nil, ErrNoGoodAddresses
 	}
 
-	if !forceDirect {
-		/////// Check backoff andnRank addresses
-		var nonBackoff bool
-		for _, a := range goodAddrs {
-			// skip addresses in back-off
-			if !s.backf.Backoff(p, a) {
-				nonBackoff = true
-			}
+	return goodAddrs, nil
+}
+
+func (s *Swarm) mergeDialContexts(a, b context.Context) context.Context {
+	dialCtx := a
+
+	if simConnect, reason := network.GetSimultaneousConnect(b); simConnect {
+		if simConnect, _ := network.GetSimultaneousConnect(a); !simConnect {
+			dialCtx = network.WithSimultaneousConnect(dialCtx, reason)
 		}
-		if !nonBackoff {
-			return nil, ErrDialBackoff
+	}
+
+	return dialCtx
+}
+
+func (s *Swarm) dialNextAddr(ctx context.Context, p peer.ID, addr ma.Multiaddr, resch chan dialResult) error {
+	// check the dial backoff
+	if forceDirect, _ := network.GetForceDirectDial(ctx); !forceDirect {
+		if s.backf.Backoff(p, addr) {
+			return ErrDialBackoff
 		}
 	}
 
-	connC, dialErr := s.dialAddrs(ctx, p, s.rankAddrs(goodAddrs))
-	if dialErr != nil {
-		logdial["error"] = dialErr.Cause.Error()
-		switch dialErr.Cause {
-		case context.Canceled, context.DeadlineExceeded:
-			// Always prefer the context errors as we rely on being
-			// able to check them.
-			//
-			// Removing this will BREAK backoff (causing us to
-			// backoff when canceling dials).
-			return nil, dialErr.Cause
+	// start the dial
+	s.limitedDial(ctx, p, addr, resch)
+
+	return nil
+}
+
+func (s *Swarm) canDial(addr ma.Multiaddr) bool {
+	t := s.TransportForDialing(addr)
+	return t != nil && t.CanDial(addr)
+}
+
+func (s *Swarm) nonProxyAddr(addr ma.Multiaddr) bool {
+	t := s.TransportForDialing(addr)
+	return !t.Proxy()
+}
+
+// ranks addresses in descending order of preference for dialing, with the following rules:
+// NonRelay > Relay
+// NonWS > WS
+// Private > Public
+// UDP > TCP
+func (s *Swarm) rankAddrs(addrs []ma.Multiaddr) []ma.Multiaddr {
+	addrTier := func(a ma.Multiaddr) (tier int) {
```

Comment:
nit: we could make this more generic by ranking according to a list of functions.

Reply:
I.e.:

```go
var rankFunctions = []func(ma.Multiaddr) bool{}

func rank(m ma.Multiaddr) int {
    rank := 0
    for i, f := range rankFunctions {
        if f(m) {
           rank |= (1<<i)
        }
    }
    return rank
}
```

Reply:
yes, we could -- let's keep this in mind for subsequent iterations; it's just 4 bits now.